bigquery/google/cloud/bigquery/table.py

```diff
@@ -1099,3 +1145,101 @@ def to_dataframe(self):
+            must be a top-level TIMESTAMP or DATE field. Its mode must be
+            NULLABLE or REQUIRED.
+        expiration_ms(int, optional):
+            Number of milliseconds for which to keep the storage for a
+            partition.
+    """
+    def __init__(self, partitioning_type, field=None, expiration_ms=None):
+        self._properties = {}
+        self.partitioning_type = partitioning_type
+        if field is not None:
+            self.field = field
+        if expiration_ms is not None:
+            self.expiration_ms = str(expiration_ms)
+
+    @property
+    def partitioning_type(self):
+        """google.cloud.bigquery.table.TimePartitioningType: The type of time
+        partitioning to use.
+        """
+        return self._properties['type']
+
+    @partitioning_type.setter
+    def partitioning_type(self, value):
+        self._properties['type'] = value
+
+    @property
+    def field(self):
+        """str: Field in the table to use for partitioning"""
+        return self._properties.get('field')
+
+    @field.setter
+    def field(self, value):
+        self._properties['field'] = value
+
+    @property
+    def expiration_ms(self):
+        """int: Number of milliseconds to keep the storage for a partition."""
+        return _helpers._int_or_none(self._properties.get('expirationMs'))
+
+    @expiration_ms.setter
+    def expiration_ms(self, value):
+        self._properties['expirationMs'] = str(value)
+
+    @classmethod
+    def from_api_repr(cls, api_repr):
+        """Return a ``TimePartitioning`` object deserialized from a ``dict``.
+
+        This method creates a new TimePartitioning instance that points to the
+        ``api_repr`` parameter as its internal properties dict. This means that
+        when a ``TimePartitioning`` instance is stored as a property of another
+        object, any changes made at the higher level will also appear here.::
+
+            >>> time_partitioning = TimePartitioning(TimePartitioningType.DAY)
+            >>> table.time_partitioning = time_partitioning
+            >>> table.time_partitioning.field = 'timecolumn'
+            >>> time_partitioning.field
+            'timecolumn'
+
+        Args:
+            api_repr (Mapping[str, str]):
+                The serialized representation of the TimePartitioning, such as
+                what is output by :meth:`to_api_repr`.
+
+        Returns:
+            google.cloud.bigquery.table.TimePartitioning:
+                The ``TimePartitioning`` object.
+        """
+        instance = cls(api_repr['type'])
+        instance._properties = api_repr
+        return instance
+
+    def to_api_repr(self):
+        """Return a dictionary representing this object.
+
+        Returns:
+            dict:
+                A dictionary representing the TimePartitioning object in
+                serialized form.
+        """
+        return self._properties
```

Comment:
Maybe add a comment why we aren't making a copy?

From chat: "Table already has some top-level properties which modify the time partitioning resource, so return a reference instead of a copy. This way any changes on the related Table properties also appear here."

Reply:
Added some text similar to what is in `from_api_repr()` docstring